### PR TITLE
fead: Add llm engine to inference crate

### PR DESCRIPTION
**Pull request by jorgeantonio21**

### CODE REVIEW

File: atoma-inference/src/serving/tokenizer.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    oneshot,
+};
+use tracing::{error, Span};
+
+/// `TokenizerRequest` - A request for the tokenizer worker tokenize
+/// the string input
+pub struct TokenizerRequest {
+    /// Input string
+    pub input: String,
+    /// `oneshot::Sender`` responsible to deliver the result of tokenization,
```
Comment (Cifko) [inline atoma-inference/src/serving/tokenizer.rs:14]:
```suggestion
    /// `oneshot::Sender` responsible to deliver the result of tokenization,
```

File: atoma-inference/src/serving/tokenizer.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    /// which includes the actual `Encoding`, together with the original input
+    /// in `String` format
+    pub sender: oneshot::Sender<Result<(Encoding, String), TokenizerError>>,
```
Comment (Cifko) [inline atoma-inference/src/serving/tokenizer.rs:17]: Maybe factor this out into a new type.

File: atoma-inference/src/serving/tokenizer.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    /// Starts the tokenizer workers
+    pub async fn start(
+        tokenizer: Tokenizer,
+        receiver: mpsc::UnboundedReceiver<TokenizerRequest>,
+        workers: usize,
+    ) -> Result<(), TokenizerError> {
+        let mut senders = Vec::with_capacity(workers);
+
+        for _ in 0..workers {
+            let tokenizer_clone = tokenizer.clone();
```
Comment (Cifko) [inline atoma-inference/src/serving/tokenizer.rs:46]:
```suggestion
            let tokenizer = tokenizer.clone();
```

File: atoma-inference/src/serving/types.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct GenerateParameters {
+    /// Generate `best_of` sequences and return the one with the highest token logprobs
+    pub best_of: Option<usize>,
+    /// Temperature is used for modeling the logits distribution
+    pub temperature: Option<f32>,
+    /// The parameter for repetition penalty. 1.0 means no penalty.
+    /// See [this paper](https://arxiv.org/pdf/1909.05858.pdf) for more details.
+    pub repetition_penalty: Option<f32>,
+    /// The parameter for frequency penalty. 1.0 means no penalty
```
Comment (Cifko) [inline atoma-inference/src/serving/types.rs:26]: no frequency penalty is 0, right?

File: atoma-inference/src/serving/tokenizer.rs (hunk 1 above)
Comment (jorgeantonio21) [reply atoma-inference/src/serving/tokenizer.rs:14]: Nice catch !

File: atoma-inference/src/serving/tokenizer.rs (hunk 2 above)
Comment (jorgeantonio21) [reply atoma-inference/src/serving/tokenizer.rs:17]: Done !

File: atoma-inference/src/serving/types.rs (hunk 4 above)
Comment (jorgeantonio21) [reply atoma-inference/src/serving/types.rs]: Oh good catch !

**Pull request merged**